src/verifier.cc

```diff
@@ -15,15 +15,143 @@
 #include "src/verifier.h"
 
 #include <cmath>
+#include <vector>
+
+#include "src/command.h"
 
 namespace amber {
 namespace {
 
-const float kEpsilon = 0.002f;
+const double kEpsilon = 0.002;
 
-bool IsFloatPixelEqualInt(float pixel, uint8_t expected) {
+bool IsFuzzyEqual(double real, double expected) {
   // TODO(jaebaek): Change kEpsilon to tolerance.
-  return std::fabs(pixel - static_cast<float>(expected) / 255.0f) < kEpsilon;
+  return std::fabs(real - expected) < kEpsilon;
+}
+
+template <typename T>
+Result CheckIntValue(ProbeSSBOCommand::Comparator comp,
+                     const uint8_t* memory,
+                     const std::vector<Value>& values) {
+  const T* ptr = reinterpret_cast<const T*>(memory);
+  switch (comp) {
+    case ProbeSSBOCommand::Comparator::kEqual:
+      for (auto v : values) {
+        if (*ptr != static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    case ProbeSSBOCommand::Comparator::kNotEqual:
+      for (auto v : values) {
+        if (*ptr == static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    case ProbeSSBOCommand::Comparator::kFuzzyEqual:
+      for (auto v : values) {
+        if (IsFuzzyEqual(static_cast<double>(*ptr),
+                         static_cast<double>(v.AsUint64())))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    case ProbeSSBOCommand::Comparator::kLess:
+      for (auto v : values) {
+        if (*ptr <= static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    case ProbeSSBOCommand::Comparator::kLessOrEqual:
+      for (auto v : values) {
+        if (*ptr < static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    case ProbeSSBOCommand::Comparator::kGreater:
+      for (auto v : values) {
+        if (*ptr >= static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    case ProbeSSBOCommand::Comparator::kGreaterOrEqual:
+      for (auto v : values) {
+        if (*ptr > static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    default:
+      return Result("Verifier::ProbeSSBO unknown comparator");
+  }
+
+  return {};
+}
+
+template <typename T>
+Result CheckFloatValue(ProbeSSBOCommand::Comparator comp,
```

Comment:
This is almost identical to the above they can probably be combined just have to change the
static_cast<T>(v.AsDouble) to something like static_cast<T>(v.IsInteger() ? v.AsUint64() : v.AsDouble()).